### PR TITLE
docs: update RPC server page to describe transport protocols

### PR DESCRIPTION
**Pull request by jmcook1186**
Expanded page to better explain differences between RPC transport protocols.

### CODE REVIEW

File: docs/_rpc/server.md [hunk 1]
```diff
@@ -3,8 +3,17 @@ title: JSON-RPC Server
 sort_key: A
 ---
 
-Geth supports all standard web3 JSON-RPC APIs. You can find documentation for
-these APIs on the [Ethereum Wiki JSON-RPC page][web3-rpc].
+Interacting with Geth requires sending requests to specific JSON_RPC API 
+endpoints. Geth supports all standard Web3 [JSON-RPC API][web3-rpc] endpoints.
```
Comment (lightclient) [inline docs/_rpc/server.md:7]:
```suggestion
methods. Geth supports all standard Web3 [JSON-RPC API][web3-rpc] endpoints. 
```

Comment (lightclient) [reply docs/_rpc/server.md:7]: Also I think "web3" is the wrong term here. Not sure if we have much of a better term though beyond "execution layer/client api". But IMO web3 has many different meanings.

File: docs/_rpc/server.md [hunk 2]
```diff
@@ -18,72 +27,197 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
-
-    geth --http
-
-By default, geth accepts connections from the loopback interface (127.0.0.1).
-The default listening port is 8545. You can customize address and port using the
-`--http.port` and `--http.addr` flags.
+## RPC
 
-    geth --http --http.port 3334
+[RPC (Remote Procedure Calls)][rpc] are programs that execute a procedure in an address
```
Comment (lightclient) [inline docs/_rpc/server.md:32]: Not sure if I would say RPCs are programs? They are a communication mechanism between programs.

File: docs/_rpc/server.md [hunk 3]
```diff
@@ -18,72 +27,197 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
+space other than their own. Address spaces can be separate processes on a single
+machine or different machines. RPC is a client-server interaction where a client sends 
+a request that includes a payload that invokes some method in the server and returns
+the result in a response to the client. The RPC has to be implemented using some transport 
+system - for Geth this transport system can be IPC, HTTP or Websockets. These are individually
+enabled using command line flags when Geth is started.
```
Comment (lightclient) [inline docs/_rpc/server.md:38]: Not sure if you want to be this granular here, but I believe IPC is active by default.

File: docs/_rpc/server.md [hunk 4]
```diff
@@ -18,72 +27,197 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
+```sh
+--http.corsdomain '*'
+```
 
 ### WebSocket Server
 
-Configuration of the WebSocket endpoint is similar to the HTTP transport. To
-enable WebSocket access, use `--ws` flag. The default WebSocket port is 8546.
+Websocket is a bidirectional transport protocol. Unlike HTTP, Websocket is *stateful*, meaning
+it maintains a connection between client and server untilit is explicitly terminated by one
```
Comment (lightclient) [inline docs/_rpc/server.md:120]:
```suggestion
it maintains a connection between client and server until it is explicitly terminated by one
```

File: docs/_rpc/server.md [hunk 5]
```diff
@@ -18,72 +27,197 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
+or other party. The persistent connection between the client and server is known as a "Websocket"
+that enables message exchange. The connection is established using a 'handshake' procedure that
```
Comment (lightclient) [inline docs/_rpc/server.md:122]: Two forms of quotes here (`"` + `'`)

File: docs/_rpc/server.md [hunk 6]
```diff
@@ -18,72 +27,197 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
+
+RPC requests to a Geth node can be made using three different transport protocols. The 
+protocols are enabled at startup using their respective flags. The right choice of transport
+protocol depends on the specific use case.
+
 
 [web3-rpc]: https://github.com/ethereum/execution-apis
 [remix]: https://remix.ethereum.org
+[rpc]: https://www.ibm.com/docs/en/aix/7.1?topic=concepts-remote-procedure-call
+
```
Comment (lightclient) [inline docs/_rpc/server.md:223]:
```suggestion
```

File: docs/_rpc/server.md [hunk 7]
```diff
@@ -18,72 +27,196 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
-
-To enable the HTTP server, use the `--http` flag.
-
-    geth --http
+## RPC
 
-By default, geth accepts connections from the loopback interface (127.0.0.1).
-The default listening port is 8545. You can customize address and port using the
-`--http.port` and `--http.addr` flags.
+[RPC (Remote Procedure Calls)][rpc] are programs that execute a procedure in an address
```
Comment (jmcook1186) [inline docs/_rpc/server.md:32]:
```suggestion
[RPC (Remote Procedure Calls)][rpc] are communication mechanisms that enable programs to execute a procedure in an address
```

File: docs/_rpc/server.md (hunk 3 above)
Comment (jmcook1186) [reply docs/_rpc/server.md:38]: ah yea good point, thanks.
```suggestion
enabled using command line flags when Geth is started (IPC is enabled by default).
```

File: docs/_rpc/server.md [hunk 8]
```diff
@@ -18,72 +27,196 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
+space other than their own. Address spaces can be separate processes on a single
+machine or different machines. RPC is a client-server interaction where a client sends 
+a request that includes a payload that invokes some method in the server and returns
+the result in a response to the client. The RPC has to be implemented using some transport 
+system - for Geth this transport system can be IPC, HTTP or Websockets. These are individually
```
Comment (jmcook1186) [inline docs/_rpc/server.md:37]:
```suggestion
system - for Geth this transport system can be IPC, HTTP or Websockets. HTTP and Websockets are individually
```

File: docs/_rpc/server.md [hunk 9]
```diff
@@ -18,72 +27,196 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
+--http.corsdomain '*'
+```
 
 ### WebSocket Server
 
-Configuration of the WebSocket endpoint is similar to the HTTP transport. To
-enable WebSocket access, use `--ws` flag. The default WebSocket port is 8546.
+Websocket is a bidirectional transport protocol. Unlike HTTP, Websocket is *stateful*, meaning
+it maintains a connection between client and server until it is explicitly terminated by one
+or other party. The persistent connection between the client and server is known as a "Websocket"
```
Comment (jmcook1186) [inline docs/_rpc/server.md:121]:
```suggestion
or other party. The persistent connection between the client and server is known as a 'Websocket'
```

File: docs/_rpc/server.md [hunk 10]
```diff
@@ -18,72 +27,197 @@ For example, the `eth_call` method resides in the `eth` namespace.
...
-namespace that isn't whitelisted. The default whitelist allows access to the "eth", "net"
-and "web3" namespaces. To overwrite the default whitelist and enable access to other APIs 
-like account management ("personal") and debugging ("debug"), they must be configured via the 
-`--http.api` flag. We do not recommend enabling such APIs over HTTP, however, 
-since access to these methods increases the attack surface.
+Regardless of the transport protocol, the RPC address determines whether the RPC methods
+are accessible from the local machine or from other machines on the same network. The default
+for Geth is to expose `localhost:8545` (equivalent to `127.0.0.1:8545`) for RPC communication.
+This default only allows other processes on the same machine to communicate with the node
+via RPC. However, using a custom non-local address (e.g. `192.168.1.321`) enables access to
```
[the remark this answers is not in the document]
Comment (jmcook1186) [reply docs/_rpc/server.md:44]: hmmm ok, thanks for the pointer. Will try to reword.

**Review by s1na — APPROVED** [review]
IMO some details about the protocols themselves don't really fit our website. People can refer to other documents for that. So I shortened those parts a bit. The table had duplicate information. Changed that too. Otherwise looks good thanks.

Btw please make sure there are no trailing spaces at the end of lines.

**Pull request merged**